### ISSUE TITLE
Python reST: Order of param's type and param's name in placeholder

### ISSUE THREAD

**Issue by fbdb** [kkoomen/vim-doge#635]
**Describe the bug**

When generating a reST docstring with Python, the parameter's name is written before the parameter's type. However, it should be the other way around: the type before the name.

I think that whoever decided the placeholders order used as a reference [this SO answer](https://stackoverflow.com/a/24385103), which points to [this blog post](http://daouzli.com/blog/docstring.html), which, indeed, put the type after the parameter's name. But I couldn't find any other references which would back this standard. [PyCharm's allow only to put the type before](https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html#81a78367), and [Sphynx's doc gives examples with the type before the name](https://www.sphinx-doc.org/en/master/usage/domains/python.html#info-field-lists).

Is this behavior intended ? Shouldn't it be the other way around ?

**To Reproduce**

After calling `:DogeGenerate` on a Python function, this is the output:

```python
def foo(bar):
    """
    [TODO:description]

    :param bar [TODO:type]: [TODO:description]
    """
    pass
```

**Expected behavior**

After calling `:DogeGenerate` on a Python function, this should be the output:

```python
def foo(bar):
    """
    [TODO:description]

    :param [TODO:type] bar: [TODO:description]
    """
    pass
```

---

NVIM v0.9.1
vim-doge-helper 4.1.0

**Comment by kkoomen** [comment]
Hi,

Thanks for discussing the issue. As noted within the doc standard within vim-doge's helper package for reST https://github.com/kkoomen/vim-doge/blob/master/helper/src/python/docs/reST.yaml#L1 it has always been following that blog post.

The reST doc standard does allow two formats:

format 1:

```py
"""
:param param1: Description of param1
:type param1: str
"""
```

format 2:

```py
"""
:param param1 str: Description of param1
"""
```

Since format 1 is the same as the Sphinx format, I made reST to be format 2 by default and let the user switch to specifically sphinx, which is format (1) if they want to. Although Sphinx does add `, defaults to <type>` in this case (which is still recommended by [PEP 257](https://peps.python.org/pep-0257/)):

```py
  def myFunc(p1, p2, p3 = ''):
    """
    [TODO:description]

    :param p1: [TODO:description]
    :type p1: [TODO:type]
    :param p2: [TODO:description]
    :type p2: [TODO:type]
    :param p3: [TODO:description], defaults to ''
    :type p3: [TODO:type], optional
    """
    pass
```

In the past 10 years I have never have come across a Python reST line that puts the type *before* the name. It has always been like this, so to answer your question: yes, it has always intentionally been this way in vim-doge.

[The PyCharm page](https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html#param-type-specification) you are referring to seems to have been [added around 2016](https://web.archive.org/web/20160912115853/http://www.jetbrains.com/help/pycharm/2016.1/using-docstrings-to-specify-types.html) if you have a look at the [Wayback Machine for that page](https://web.archive.org/web/20161015000000*/https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html). At that time they did refer to another internal page [Python Integrated Tools](https://web.archive.org/web/20161112054421/http://www.jetbrains.com/help/pycharm/2016.1/python-integrated-tools.html) which lists the doc standards but the statement about the order [seems to have been added around 2020](https://web.archive.org/web/20200104004332/https://www.jetbrains.com/help/pycharm/using-docstrings-to-specify-types.html). Therefore, my personal guess is that they have recently made a personal preference by slightly re-ordering the reST structure. Maybe because in most standards they always do the type and then the name. However, for reST it doesn't matter and when I test both orders in PyCharm, nothing is mentioned on this.

**Comment by fbdb** [comment]
Hi, Thank you for taking the time to answer.

---

> ... and when I test both orders in PyCharm, nothing is mentioned on this.

Strangely, on my side, whenever I open a snippet of code in which a doctring has been generated with vim-doge, PyCharm "complains" that the param's docstring is missing:

![image](https://github.com/kkoomen/vim-doge/assets/50629818/e04e3a85-3a6a-469d-a204-3d9575e46c09)

By hovering the param `bar`, a pop up appears, saying *Missing parameter bar in docstring*. And by hovering the `str` in the *:param* line, it says *Function 'foo' does not have a parameter 'str'*.
I tried to search a bit if there is a hidden setting in PyCharm for configuring the order between the type and the name, but with no luck... Maybe your PyCharm has a different setting than mine somewhere ...?

---

> since the blog post has been leading in the reST standard for many more years.

I understand your point of view. However, I wanted to ask, what's your opinion about the second link that I provided in my first message, the [Sphynx's doc which gives examples with the type before the name](https://www.sphinx-doc.org/en/master/usage/domains/python.html#info-field-lists) ? It seems quite "official", or at least more than the blog post that you've used as a reference.

---

I hope I don't sound rude, I just want to understand.
